// uxcore-crumb/build/prototype.js
'use strict';

var React = require('react');

var _require = require('engine');

var Bundle = _require.Bundle;

var _require2 = require('engine-utils');

var ListSetter = _require2.ListSetter;


module.exports = Bundle.createPrototype({
    title: "面包屑",
    category: "*",
    icon: require("./logo.svg"), // todo: require("./logo.svg"),
    componentName: "Crumb",
    canHovering: true,
    canSelecting: true,
    canDraging: true,
    isInline: false,
    isContainer: true,
    canDropto: true,
    canDroping: false,
    initialChildren: [{
        componentName: "CrumbItem",
        props: {
            children: "标签A",
            href: "http://www.taobao.com"
        }
    }, {
        componentName: "CrumbItem",
        props: {
            children: "标签B"
        }
    }],
    configure: [{
        name: "items",
        title: "面包屑项",
        fieldStyle: "block",
        setter: React.createElement(ListSetter, { primaryKey: '', titleField: 'children', addData: { children: "面包屑项" } }),
        ignore: true,
        accessor: function accessor() {
            var node = this.getNode();
            var children = node.getChildren();
            return children.map(function (child) {
                var childText = child.getPropValue('children');
                var href = child.getPropValue('href');
                var data = {
                    content: childText,
                    href: href
                };
                return data;
            });
        },
        mutator: function mutator(value, hotValue) {
            var node = this.getNode();
            node.mergeChildren(function (child, index) {
                return true;
            }, function (children) {
                return value.map(function (item) {
                    return { componentName: "CrumbItem", props: item };
                });
            }, function (child1, child2) {
                return -1;
            });
        }
    }]
});
