// uxcore-table/build/Cell/Radio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

/**
 * A radio field
 */

var Const = require('uxcore-const');
var React = require('react');
var ReactDOM = require('react-dom');

var Radio = function (_React$Component) {
    _inherits(Radio, _React$Component);

    function Radio(props) {
        _classCallCheck(this, Radio);

        return _possibleConstructorReturn(this, _React$Component.call(this, props));
    }

    Radio.prototype.handleChange = function handleChange(e) {
        this.props.onChange.apply(null, [e]);
    };

    Radio.prototype.getValue = function getValue() {
        return this.refs.radio.checked;
    };

    Radio.prototype.render = function render() {

        var props = this.props;

        if (props.mode !== Const.MODE.VIEW) {
            var renderProps = {
                className: "kuma-checkbox",
                checked: this.props.checked,
                onChange: this.handleChange.bind(this)
            };
            if (!!props.disable) {
                renderProps.disabled = true;
            }
            return React.createElement(
                'label',
                { className: 'kuma-uxtable-row-selector' },
                React.createElement('input', _extends({ type: 'radio', ref: 'radio' }, renderProps)),
                React.createElement('s', null)
            );
        } else {

            var _renderProps = {
                className: "kuma-checkbox",
                checked: this.props.checked,
                disabled: true
            };
            return React.createElement(
                'label',
                { className: 'kuma-uxtable-row-selector' },
                React.createElement('input', _extends({ type: 'radio', ref: 'radio' }, _renderProps)),
                React.createElement('s', null)
            );
        }
    };

    return Radio;
}(React.Component);

;

Radio.propTypes = {};

Radio.defaultProps = {};

exports["default"] = Radio;
module.exports = exports['default'];
