// uxcore-steps/build/Steps.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Steps Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author vincent.bian
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2015, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var Steps = function (_React$Component) {
    _inherits(Steps, _React$Component);

    function Steps(props) {
        _classCallCheck(this, Steps);

        var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

        _this.state = {
            init: false,
            tailWidth: 0
        };
        _this._previousStepsWidth = 0;
        _this._itemsWidth = [];
        return _this;
    }

    Steps.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
        if (nextProps.children.length !== this.props.children.length) {
            if (this.props.direction === 'vertical') {
                return;
            }
            var $dom = ReactDOM.findDOMNode(this);
            var len = nextProps.children.length - 1;
            this._itemsWidth = new Array(len + 1);

            var i = void 0;
            for (i = 0; i <= len - 1; i++) {
                this._itemsWidth[i] = nextProps.maxDescriptionWidth;
            }
            this._itemsWidth[i] = nextProps.maxDescriptionWidth;
            this._update(nextProps);
        }
    };

    Steps.prototype.componentDidMount = function componentDidMount() {
        var _this2 = this;

        if (this.props.direction === 'vertical') {
            return;
        }
        var $dom = ReactDOM.findDOMNode(this);
        var len = $dom.children.length - 1;
        this._itemsWidth = new Array(len + 1);

        var i = void 0;
        for (i = 0; i <= len - 1; i++) {
            this._itemsWidth[i] = this.props.maxDescriptionWidth;
        }
        this._itemsWidth[i] = this.props.maxDescriptionWidth;
        this._previousStepsWidth = Math.floor(ReactDOM.findDOMNode(this).offsetWidth);
        this._update();

        /*
         * 把最后一个元素设置为absolute，是为了防止动态添加元素后滚动条出现导致的布局问题。
         * 未来不考虑ie8一类的浏览器后，会采用纯css来避免各种问题。
         */
        $dom.children[len].style.position = 'absolute';

        /*
         * 下面的代码是为了兼容window系统下滚动条出现后会占用宽度的问题。
         * componentDidMount时滚动条还不一定出现了，这时候获取的宽度可能不是最终宽度。
         * 对于滚动条不占用宽度的浏览器，下面的代码也不二次render，_resize里面会判断要不要更新。
         */
        setTimeout(function () {
            _this2._resize();
        });

        this._resizeBind = this._resize.bind(this);

        if (window.attachEvent) {
            window.attachEvent('onresize', this._resizeBind);
        } else {
            window.addEventListener('resize', this._resizeBind);
        }
    };

    Steps.prototype.componentDidUpdate = function componentDidUpdate() {
        this._resize();
        var $dom = ReactDOM.findDOMNode(this);

        var len = $dom.children.length - 1;
        /*
         * 把最后一个元素设置为absolute，是为了防止动态添加元素后滚动条出现导致的布局问题。
         * 未来不考虑ie8一类的浏览器后，会采用纯css来避免各种问题。
         */
        for (var i = 0; i <= len; i++) {
            $dom.children[i].style.position = 'relative';
        }
        $dom.children[len].style.position = 'absolute';
    };

    Steps.prototype.componentWillUnmount = function componentWillUnmount() {
        if (this.props.direction === 'vertical') {
            return;
        }
        if (window.attachEvent) {
            window.detachEvent('onresize', this._resizeBind);
        } else {
            window.removeEventListener('resize', this._resizeBind);
        }
    };

    Steps.prototype._resize = function _resize() {
        var w = Math.floor(ReactDOM.findDOMNode(this).offsetWidth);
        if (this._previousStepsWidth === w) {
            return;
        }
        this._previousStepsWidth = w;
        this._update();
    };

    Steps.prototype._update = function _update() {
        var props = arguments.length <= 0 || arguments[0] === undefined ? this.props : arguments[0];

        var len = props.children.length - 1;
        var tw = this._itemsWidth.reduce(function (prev, w) {
            return prev + w;
        }, 0);
        var dw = Math.floor((this._previousStepsWidth - tw) / len) - 1;
        if (dw <= 0) {
            return;
        }
        this.setState({
            init: true,
            tailWidth: dw
        });
    };

    Steps.prototype.render = function render() {
        var _this3 = this;

        var props = this.props;
        var prefixCls = props.prefixCls;
        var className = props.className;
        var children = props.children;
        var maxDescriptionWidth = props.maxDescriptionWidth;
        var iconPrefix = props.iconPrefix;
        var size = props.size;
        var direction = props.direction;
        var showIcon = props.showIcon;
        var current = props.current;
        var type = props.type;

        var len = children.length - 1;
        var iws = this._itemsWidth;
        var clsName = prefixCls;
        if (direction === 'vertical') {
            clsName += ' ' + prefixCls + '-vertical ' + className;
        } else {
            clsName += ' ' + prefixCls + '-type-' + type + ' ' + className;
        }
        if (!showIcon) {
            clsName += ' ' + prefixCls + '-noicon';
        }

        return _react2["default"].createElement(
            'div',
            { className: clsName },
            _react2["default"].Children.map(children, function (ele, idx) {
                var np = {
                    stepNumber: showIcon ? (idx + 1).toString() : '',
                    stepLast: idx === len,
                    tailWidth: iws.length === 0 || idx === len ? 'auto' : iws[idx] + _this3.state.tailWidth,
                    prefixCls: prefixCls,
                    iconPrefix: iconPrefix,
                    maxDescriptionWidth: maxDescriptionWidth
                };
                if (!ele.props.status) {
                    np.status = idx === current ? 'process' : idx < current ? 'finish' : 'wait';
                }
                return _react2["default"].cloneElement(ele, np);
            }, this)
        );
    };

    return Steps;
}(_react2["default"].Component);

Steps.defaultProps = {
    prefixCls: 'kuma-step',
    className: '',
    iconPrefix: '',
    maxDescriptionWidth: 100,
    current: 0,
    direction: '',
    showIcon: true,
    type: 'default'
};

// http://facebook.github.io/react/docs/reusable-components.html
Steps.propTypes = {
    prefixCls: _react2["default"].PropTypes.string,
    className: _react2["default"].PropTypes.string,
    iconPrefix: _react2["default"].PropTypes.string,
    maxDescriptionWidth: _react2["default"].PropTypes.number,
    current: _react2["default"].PropTypes.number,
    direction: _react2["default"].PropTypes.string,
    showIcon: _react2["default"].PropTypes.bool,
    type: _react2["default"].PropTypes.oneOf(['default', 'title-on-top', 'long-desc'])
};

Steps.displayName = "Steps";

exports["default"] = Steps;
module.exports = exports['default'];
