// uxcore-tabs/build/Tabs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _rcTabs = require('rc-tabs');

var _rcTabs2 = _interopRequireDefault(_rcTabs);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var prefixCls = 'kuma-tab';
var TYPESUFFIX = {
	large: 'lg',
	small: 'sm',
	filter: 'filter',
	brick: 'brick'
};

var Tabs = function (_RcTabs) {
	_inherits(Tabs, _RcTabs);

	function Tabs() {
		_classCallCheck(this, Tabs);

		return _possibleConstructorReturn(this, _RcTabs.apply(this, arguments));
	}

	Tabs.prototype.render = function render() {
		var _classnames;

		var props = this.props;
		return _react2["default"].createElement(_rcTabs2["default"], _extends({}, props, { className: (0, _classnames3["default"])((_classnames = {}, _defineProperty(_classnames, prefixCls + '-' + TYPESUFFIX[props.type], true), _defineProperty(_classnames, props.className, !!props.className), _classnames)) }));
	};

	return Tabs;
}(_rcTabs2["default"]);

Tabs.displayName = 'uxcore-tabs';
Tabs.defaultProps = (0, _objectAssign2["default"])(_rcTabs2["default"].defaultProps, {
	prefixCls: prefixCls,
	type: 'large'
});
Tabs.TabPane = _rcTabs2["default"].TabPane;

exports["default"] = Tabs;
module.exports = exports['default'];
