// uxcore-button/build/prototype.js
'use strict';

var React = require('react');

var _require = require('engine');

var Bundle = _require.Bundle;

var _require2 = require('engine-utils');

var BoolSetter = _require2.BoolSetter;
var TextSetter = _require2.TextSetter;
var ChoiceSetter = _require2.ChoiceSetter;
var JsonSetter = _require2.JsonSetter;
var NumberSetter = _require2.NumberSetter;


module.exports = Bundle.createPrototype({
    title: "按钮",
    category: "*",
    icon: require("./logo.svg"), // todo: require("./logo.svg"),
    componentName: "Button",
    canHovering: true,
    canSelecting: true,
    canDraging: true,
    isInline: true,
    isContainer: false,
    canDropto: true,
    canDroping: false,
    configure: [{
        name: "size",
        title: "尺寸",
        defaultValue: 'medium',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(ChoiceSetter, { options: [{ value: 'small' }, { value: 'medium' }, { value: 'large' }] })
    }, {
        name: "style",
        title: "样式",
        defaultValue: {},
        required: false,
        fieldStyle: "accordion",
        fieldCollapsed: false,
        setter: React.createElement(JsonSetter, null)
    }, {
        name: "type",
        title: "类型",
        defaultValue: 'primary',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(ChoiceSetter, { options: [{ value: 'primary' }, { value: 'secondary' }, { value: 'outline' }, { value: 'disabled' }] })
    }, {
        name: "children",
        title: "内容",
        defaultValue: 'Button',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, { multiline: true, rows: 2 })
    }]
});
