// uxcore-mention/build/components/Mention.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Panel = require('./Panel');

var _Panel2 = _interopRequireDefault(_Panel);

var _reactMixin = require('react-mixin');

var _reactMixin2 = _interopRequireDefault(_reactMixin);

var _keycode = require('../utils/keycode');

var _mentionMixin = require('./mentionMixin');

var _mentionMixin2 = _interopRequireDefault(_mentionMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author: vincent.bian
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var __matchTimer = void 0;

var Mention = function (_React$Component) {
  _inherits(Mention, _React$Component);

  function Mention(props) {
    _classCallCheck(this, Mention);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      mentionList: [],
      cursorPosition: {
        x: 0,
        y: 0
      },
      panelVisible: false,
      panelIdx: 0
    };
    return _this;
  }

  Mention.prototype.componentDidMount = function componentDidMount() {
    this.activeEditor = null;
  };

  Mention.prototype.onFocus = function onFocus(editorInstance) {
    this.activeEditor = editorInstance;
  };

  Mention.prototype.setPanelPos = function setPanelPos(pos) {
    var position = {
      x: pos.x,
      y: pos.y
    };
    this.setState({
      cursorPosition: position
    });
  };

  Mention.prototype.selectItem = function selectItem(data) {
    this.setState({
      mentionList: []
    });
    this.activeEditor.insertMentionData(data);
  };

  Mention.prototype.render = function render() {
    var _this2 = this;

    var panelPosition = {
      left: this.state.cursorPosition.x,
      top: this.state.cursorPosition.y
    };
    var _props = this.props;
    var prefixCls = _props.prefixCls;
    var readOnly = _props.readOnly;
    var onChange = _props.onChange;
    var children = _props.children;
    var panelFormatter = _props.panelFormatter;

    return _react2["default"].createElement(
      'div',
      { onKeyUp: this.onPanelKeyup.bind(this) },
      _react2["default"].Children.map(children, function (Comp) {
        return _react2["default"].cloneElement(Comp, {
          prefixCls: prefixCls,
          panelVisible: _this2.state.panelVisible,
          matcher: _this2.runMatcher.bind(_this2),
          setCursorPos: _this2.setPanelPos.bind(_this2),
          onChange: onChange,
          onFocus: _this2.onFocus.bind(_this2)
        });
      }),
      _react2["default"].createElement(_Panel2["default"], {
        prefixCls: prefixCls,
        visible: this.state.panelVisible,
        idx: this.state.panelIdx,
        list: this.state.mentionList,
        onSelect: this.selectItem.bind(this),
        formatter: panelFormatter,
        style: panelPosition
      })
    );
  };

  return Mention;
}(_react2["default"].Component);

(0, _reactMixin2["default"])(Mention.prototype, _mentionMixin2["default"]);

Mention.propType = {
  prefixCls: _react2["default"].PropTypes.string,
  source: _react2["default"].PropTypes.oneOfType([_react2["default"].PropTypes.array, _react2["default"].PropTypes.func]),
  delay: _react2["default"].PropTypes.number,
  matchRange: _react2["default"].PropTypes.arrayOf(_react2["default"].PropTypes.number),
  formatter: _react2["default"].PropTypes.func,
  panelFormatter: _react2["default"].PropTypes.func,
  onChange: _react2["default"].PropTypes.func
};
Mention.defaultProps = {
  prefixCls: 'kuma-mention',
  source: [],
  delay: 100,
  matchRange: [2, 8],
  formatter: function formatter(data) {
    return data;
  },
  panelFormatter: function panelFormatter(data) {
    return '' + data.text;
  },
  onChange: function onChange(e, value) {}
};

exports["default"] = Mention;
module.exports = exports['default'];
