// uxcore-mention/build/components/TinymceMention.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _reactMixin = require('react-mixin');

var _reactMixin2 = _interopRequireDefault(_reactMixin);

var _Panel = require('./Panel');

var _Panel2 = _interopRequireDefault(_Panel);

var _baseEditor = require('../editor-components/baseEditor');

var _baseEditor2 = _interopRequireDefault(_baseEditor);

var _keycode = require('../utils/keycode');

var _util = require('../utils/util');

var _mentionMixin = require('./mentionMixin');

var _mentionMixin2 = _interopRequireDefault(_mentionMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

function pluginInitialized() {
  var ed = window.tinymce.activeEditor;
  var plugins = ed && ed.plugins;
  var mention = plugins && plugins.mention;
  return !!mention;
}

var TinymceMention = function (_BaseEditor) {
  _inherits(TinymceMention, _BaseEditor);

  function TinymceMention(props) {
    _classCallCheck(this, TinymceMention);

    var _this = _possibleConstructorReturn(this, _BaseEditor.call(this, props));

    _this.state = {
      mentionList: [],
      cursorPosition: {
        x: 0,
        y: 0
      },
      panelVisible: false,
      panelIdx: 0
    };
    return _this;
  }

  TinymceMention.prototype.componentWillMount = function componentWillMount() {
    var _this2 = this;

    if (!pluginInitialized()) {
      tinymce.PluginManager.add('mention', function (activeEditor) {
        _this2.editor = activeEditor;
        activeEditor.on('keydown', function (e) {
          _this2.onKeydown(e);
        });
        activeEditor.on('keyup', function (e) {
          _this2.onKeyup(e);
          _this2.onPanelKeyup(e);
        });
      });
    }
  };

  TinymceMention.prototype.componentDidMount = function componentDidMount() {
    this.STORE = {};
    var container = this.refs.target.parentNode;
    var mceNode = document.createElement('div');
    this.mceNode = mceNode;
    _reactDom2["default"].render(_react2["default"].createElement(
      'div',
      null,
      _react2["default"].Children.map(this.props.children, function (Comp) {
        var cp = void 0;
        if (Comp.type.name === 'Tinymce' || Comp.type.displayName === 'Tinymce') {
          cp = {
            config: {
              plugins: ['mention']
            }
          };
        }
        return _react2["default"].cloneElement(Comp, cp);
      })
    ), mceNode);
    container.insertBefore(mceNode, this.refs.target);
  };

  TinymceMention.prototype.componentWillUnmount = function componentWillUnmount() {
    var container = this.refs.target.parentNode;
    container.removeChild(this.mceNode);
  };

  TinymceMention.prototype.onKeydown = function onKeydown(e) {
    var panelVisible = this.state.panelVisible;

    switch (e.keyCode) {
      case _keycode.KEYCODE.UP:
      case _keycode.KEYCODE.DOWN:
        if (panelVisible) {
          e.preventDefault();
        }
        break;
      case _keycode.KEYCODE.ENTER:
        if (panelVisible) {
          e.preventDefault();
        }
        break;
      default:
        break;
    }
  };

  TinymceMention.prototype.onKeyup = function onKeyup(e) {
    var panelVisible = this.state.panelVisible;

    switch (e.keyCode) {
      case _keycode.KEYCODE.UP:
      case _keycode.KEYCODE.DOWN:
        if (panelVisible) {
          e.preventDefault();
        }
        break;
      case _keycode.KEYCODE.ENTER:
        break;
      default:
        this.handleDefaultKeyup();
        break;
    }
  };

  TinymceMention.prototype.handleDefaultKeyup = function handleDefaultKeyup() {
    var delimiter = this.props.delimiter;

    var sel = this.editor.selection;
    var range = sel.getRng(true);
    if (range.commonAncestorContainer.nodeType === 3) {
      var cloneRange = range.cloneRange();
      cloneRange.setStart(range.commonAncestorContainer, 0);
      var originStr = cloneRange.toString();
      var str = (0, _util.parseStrByDelimiter)(originStr, delimiter);
      this.runMatcher(str);
      if (str) {
        if ('createRange' in document) {
          range.setStart(range.commonAncestorContainer, originStr.length - str.length - 1);
          var rect = range.getBoundingClientRect();
          this.setPanelPos({
            left: rect.right,
            top: rect.bottom
          });
          sel.setRng(range);
          // save range position
          this.STORE.bookmark = sel.getBookmark(3);
        } else {
          // IE8
          var internalRange = sel.getRng();
          this.setPanelPos({
            left: internalRange.boundingLeft,
            top: internalRange.boundingTop
          });
          this.STORE.bookmark = str.length + 1;
        }
        sel.collapse();
      }
    }
    // debugger;
  };

  TinymceMention.prototype.insert = function insert(mentionContent) {
    var insertMode = this.props.insertMode;

    switch (insertMode) {
      case 'TEXT_NODE':
        this.insertWithTextNode(mentionContent);
        break;
      case 'ELEMENT_NODE':
      default:
        this.insertWithElementNode(mentionContent);
        break;
    }
  };

  TinymceMention.prototype.insertWithElementNode = function insertWithElementNode(mentionContent) {
    var sel = this.editor.selection;
    if (this.STORE.bookmark) {
      sel.moveToBookmark(this.STORE.bookmark);
    }
    var range = sel.getRng(true);
    var mentionNode = document.createElement('input');
    mentionNode.setAttribute('type', 'button');
    mentionNode.setAttribute('tabindex', '-1');
    mentionNode.className = this.props.prefixCls + '-node';
    mentionNode.value = mentionContent;
    range.deleteContents();
    sel.setNode(mentionNode);
    range.collapse();
    this.editor.focus();
  };

  TinymceMention.prototype.insertWithTextNode = function insertWithTextNode(mentionContent) {
    var sel = this.editor.selection;
    if (this.STORE.bookmark) {
      if ('createRange' in document) {
        sel.moveToBookmark(this.STORE.bookmark);
        var range = sel.getRng(true);
        range.deleteContents();
        sel.setContent(mentionContent);
        range.collapse();
        this.editor.focus();
      } else {
        var internalRange = sel.getRng();
        internalRange.moveStart('character', -this.STORE.bookmark);
        internalRange.pasteHTML(mentionContent);
      }
    }
  };

  TinymceMention.prototype.setPanelPos = function setPanelPos(pos) {
    var editorOffset = this.editor.contentAreaContainer.getBoundingClientRect();
    // const offset = getScrollOffset();
    var position = {
      x: pos.left + editorOffset.left,
      y: pos.top + editorOffset.top
    };
    this.setState({
      cursorPosition: position
    });
  };

  TinymceMention.prototype.selectItem = function selectItem(data) {
    this.insertMentionData(data);
    this.setState({
      mentionList: []
    });
  };

  TinymceMention.prototype.render = function render() {
    var panelPosition = {
      left: this.state.cursorPosition.x,
      top: this.state.cursorPosition.y
    };
    var _props = this.props;
    var prefixCls = _props.prefixCls;
    var panelFormatter = _props.panelFormatter;

    return _react2["default"].createElement(
      'div',
      { ref: 'target' },
      _react2["default"].createElement(_Panel2["default"], {
        prefixCls: prefixCls,
        visible: this.state.panelVisible,
        idx: this.state.panelIdx,
        list: this.state.mentionList,
        onSelect: this.selectItem.bind(this),
        formatter: panelFormatter,
        style: panelPosition
      })
    );
  };

  return TinymceMention;
}(_baseEditor2["default"]);

(0, _reactMixin2["default"])(TinymceMention.prototype, _mentionMixin2["default"]);

TinymceMention.propType = {
  prefixCls: _react2["default"].PropTypes.string,
  source: _react2["default"].PropTypes.oneOfType([_react2["default"].PropTypes.array, _react2["default"].PropTypes.func]),
  delay: _react2["default"].PropTypes.number,
  matchRange: _react2["default"].PropTypes.arrayOf(_react2["default"].PropTypes.number),
  formatter: _react2["default"].PropTypes.func,
  mentionFormatter: _react2["default"].PropTypes.func,
  panelFormatter: _react2["default"].PropTypes.func,
  onChange: _react2["default"].PropTypes.func,
  onAdd: _react2["default"].PropTypes.func,
  insertMode: _react2["default"].PropTypes.oneOf(['ELEMENT_NODE', 'TEXT_NODE'])
};
TinymceMention.defaultProps = {
  prefixCls: 'kuma-mention',
  source: [],
  delay: 100,
  matchRange: [2, 8],
  formatter: function formatter(data) {
    return data;
  },
  mentionFormatter: function mentionFormatter(data) {
    return '@' + data.text;
  },
  panelFormatter: function panelFormatter(data) {
    return '' + data.text;
  },
  onChange: function onChange(e, value) {},
  onAdd: function onAdd() {},
  insertMode: 'ELEMENT_NODE'
};

exports["default"] = TinymceMention;
module.exports = exports['default'];
