// uxcore-dialog/build/Dialog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _rcDialog = require('rc-dialog');

var _rcDialog2 = _interopRequireDefault(_rcDialog);

var _confirm = require('./confirm');

var _confirm2 = _interopRequireDefault(_confirm);

var _uxcoreButton = require('uxcore-button');

var _uxcoreButton2 = _interopRequireDefault(_uxcoreButton);

var _i18n = require('./i18n');

var _i18n2 = _interopRequireDefault(_i18n);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author: vincent
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @date: 15/5/19
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


function noop() {}

var Dialog = function (_React$Component) {
    _inherits(Dialog, _React$Component);

    function Dialog(props) {
        _classCallCheck(this, Dialog);

        var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

        _this.state = {
            confirmLoading: false,
            visible: props.visible
        };
        return _this;
    }

    Dialog.prototype.handleCancel = function handleCancel() {
        this.props.onCancel();
    };

    Dialog.prototype.handleOk = function handleOk() {
        this.props.onOk();
    };

    Dialog.prototype.render = function render() {
        var props = this.props;
        var transitionName = props.transitionName;

        var otherProps = _objectWithoutProperties(props, ['transitionName']);

        var locale = _i18n2["default"][props.locale];

        if ((typeof document === 'undefined' ? 'undefined' : _typeof(document)) == "object") {
            if (/Safari/.test(navigator.userAgent) && /Apple Computer/.test(navigator.vendor)) {
                // safari animation bug when using threeFallV
                transitionName = 'slideDown';
            }
        }
        var defaultFooter = [_react2["default"].createElement(
            _uxcoreButton2["default"],
            { key: 'confirm',
                type: 'primary',
                size: 'medium',
                loading: props.confirmLoading,
                onClick: this.handleOk.bind(this) },
            locale['ok']
        ), _react2["default"].createElement(
            _uxcoreButton2["default"],
            { key: 'cancel',
                type: 'secondary',
                size: 'medium',
                onClick: this.handleCancel.bind(this) },
            locale['cancel']
        )];
        var footer = props.footer || defaultFooter;
        var className = void 0;
        if (!props.title) {
            className = props.className + ' ' + props.prefixCls + '-noheader';
        } else {
            className = props.className;
        }
        return _react2["default"].createElement(_rcDialog2["default"], _extends({
            onClose: this.handleCancel.bind(this),
            footer: footer
        }, props, {
            transitionName: transitionName,
            className: className,
            visible: props.visible }));
    };

    return Dialog;
}(_react2["default"].Component);

exports["default"] = Dialog;


Dialog.defaultProps = {
    prefixCls: 'kuma-dlg',
    className: '',
    wrapClassName: '',
    onOk: noop,
    locale: 'zh-cn',
    onCancel: noop,
    width: 520,
    transitionName: 'threeFallV',
    maskTransitionName: 'fade',
    confirmLoading: false,
    visible: false,
    closable: true,
    maskClosable: false,
    title: ''
};

Dialog.info = function (props) {
    props.iconClassName = 'kuma-icon-information';
    props.okCancel = false;
    return (0, _confirm2["default"])(props);
};

Dialog.success = function (props) {
    props.iconClassName = 'kuma-icon-success';
    props.okCancel = false;
    return (0, _confirm2["default"])(props);
};

Dialog.error = function (props) {
    props.iconClassName = 'kuma-icon-error';
    props.okCancel = false;
    return (0, _confirm2["default"])(props);
};

Dialog.confirm = function (props) {
    props.okCancel = true;
    return (0, _confirm2["default"])(props);
};
module.exports = exports['default'];
