// uxcore-form-row/build/prototype.js
'use strict';

var React = require('react');

var _require = require('engine');

var Bundle = _require.Bundle;

var _require2 = require('engine-utils');

var BoolSetter = _require2.BoolSetter;
var TextSetter = _require2.TextSetter;
var ChoiceSetter = _require2.ChoiceSetter;
var JsonSetter = _require2.JsonSetter;
var NumberSetter = _require2.NumberSetter;


module.exports = Bundle.createPrototype({
    title: "行",
    category: "表单",
    icon: require("./logo.svg"), // todo: require("./logo.svg"),
    componentName: "FormRow",
    canHovering: true,
    canSelecting: true,
    canDraging: true,
    isInline: false,
    isContainer: true,
    canDropto: "Form",
    canDroping: function canDroping(placement) {
        var name = placement.getComponentName();
        return (/FormField/.test(name)
        );
    },
    configure: [{
        name: "totalFlex",
        title: "弹性比例总和",
        defaultValue: null,
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(NumberSetter, null)
    }]
});
