// uxcore-calendar/build/YearCalendar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _YearPanel = require('rc-calendar/lib/year/YearPanel');

var _YearPanel2 = _interopRequireDefault(_YearPanel);

var _CalendarMixin = require('rc-calendar/lib/mixin/CalendarMixin');

var _CalendarMixin2 = _interopRequireDefault(_CalendarMixin);

var _CommonMixin = require('rc-calendar/lib/mixin/CommonMixin');

var _CommonMixin2 = _interopRequireDefault(_CommonMixin);

var _KeyCode = require('rc-util/lib/KeyCode');

var _KeyCode2 = _interopRequireDefault(_KeyCode);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var YearCalendar = _react2["default"].createClass({
    displayName: 'YearCalendar',

    mixins: [_CalendarMixin2["default"], _CommonMixin2["default"]],

    onKeyDown: function onKeyDown(e) {
        var keyCode = e.keyCode;
        var ctrlKey = e.ctrlKey || e.metaKey;
        var stateValue = this.state.value;
        var value = stateValue;
        switch (keyCode) {
            case _KeyCode2["default"].DOWN:
                value = stateValue.clone();
                value.addMonth(3);
                break;
            case _KeyCode2["default"].UP:
                value = stateValue.clone();
                value.addMonth(-3);
                break;
            case _KeyCode2["default"].LEFT:
                value = stateValue.clone();
                if (ctrlKey) {
                    value.addYear(-1);
                } else {
                    value.addMonth(-1);
                }
                break;
            case _KeyCode2["default"].RIGHT:
                value = stateValue.clone();
                if (ctrlKey) {
                    value.addYear(1);
                } else {
                    value.addMonth(1);
                }
                break;
            case _KeyCode2["default"].ENTER:
                this.onSelect(stateValue);
                e.preventDefault();
                return 1;
            default:
                return undefined;
        }
        if (value !== stateValue) {
            this.setValue(value);
            e.preventDefault();
            return 1;
        }
    },
    render: function render() {
        var props = this.props;
        var children = _react2["default"].createElement(_YearPanel2["default"], { locale: props.locale,
            disabledDate: props.disabledDate,
            value: this.state.value,
            rootPrefixCls: props.prefixCls,
            onChange: this.setValue,
            onSelect: this.onSelect });
        return this.renderRoot({
            children: children,
            className: props.prefixCls + '-year-calendar-container'

        });
    }
});

exports["default"] = YearCalendar;
module.exports = exports['default'];
