// uxcore-other-form-field/build/prototype.js
'use strict';

var React = require('react');

var _require = require('engine');

var Bundle = _require.Bundle;

var _require2 = require('engine-utils');

var BoolSetter = _require2.BoolSetter;
var TextSetter = _require2.TextSetter;
var ChoiceSetter = _require2.ChoiceSetter;
var JsonSetter = _require2.JsonSetter;
var NumberSetter = _require2.NumberSetter;


module.exports = Bundle.createPrototype({
    title: "其他",
    category: "表单",
    icon: require("./logo.svg"), // todo: require("./logo.svg"),
    componentName: "OtherFormField",
    canHovering: true,
    canSelecting: true,
    canDraging: true,
    isInline: true,
    isContainer: false,
    canDropto: function canDropto(container) {
        var name = container.getComponentName();
        return (/Form|FormRow/.test(name)
        );
    },
    canDroping: false,
    configure: [{
        name: "jsxflex",
        title: "弹性比例",
        defaultValue: 1,
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(NumberSetter, null)
    }]
});
